File: src/golang/integrations.go
package oresmiddleware

import (
	"context"
	"net/http"
	"sync"
	"time"
)

type AuthDecision struct {
	UserID   string
	TenantID string
	Claims   map[string]string
}

type AuthVerifier interface {
	Verify(context.Context, *http.Request, RequestContext) (AuthDecision, error)
}
type TestIdentityResolver interface {
	Resolve(context.Context, *http.Request, RequestContext) (AuthDecision, error)
}
type IPAuthorizer interface {
	Allow(context.Context, *http.Request, RequestContext) (bool, error)
}
type SyncObserver interface {
	Observe(context.Context, RequestContext, *http.Request, int, time.Duration) error
}
type TelemetrySink interface {
	Started(context.Context, RequestContext, *http.Request)
	Finished(context.Context, RequestContext, *http.Request, int, time.Duration)
}
type SchemaCapture interface {
	Capture(context.Context, *http.Request, int, http.Header, []byte) error
}
type RateLimiter interface {
	Allow(context.Context, string, int, float64) (bool, error)
}
type IdempotencyStore interface {
	Get(context.Context, string) (StoredResponse, bool, error)
	Put(context.Context, string, StoredResponse, time.Duration) error
}

type StoredResponse struct {
	Status    int
	Header    http.Header
	Body      []byte
	ExpiresAt time.Time
}

type Dependencies struct {
	AuthVerifier             AuthVerifier
	TestIdentity             TestIdentityResolver
	IPAuthorizer             IPAuthorizer
	SyncObserver             SyncObserver
	Telemetry                TelemetrySink
	SchemaCapture            SchemaCapture
	OperationFailureReporter OperationFailureReporter
	RateLimiter              RateLimiter
	IdempotencyStore         IdempotencyStore
	Now                      func() time.Time
	RandomFloat64            func() float64
}

type bucket struct {
	tokens  float64
	updated time.Time
}
type MemoryTokenBucket struct {
	mu      sync.Mutex
	buckets map[string]bucket
	now     func() time.Time
}

func NewMemoryTokenBucket(now func() time.Time) *MemoryTokenBucket {
	return &MemoryTokenBucket{buckets: make(map[string]bucket), now: now}
}

func (m *MemoryTokenBucket) Allow(_ context.Context, key string, capacity int, refill float64) (bool, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	now := m.now()
	value, ok := m.buckets[key]
	if !ok {
		value = bucket{tokens: float64(capacity), updated: now}
	}
	value.tokens += now.Sub(value.updated).Seconds() * refill
	if value.tokens > float64(capacity) {
		value.tokens = float64(capacity)
	}
	value.updated = now
	allowed := value.tokens >= 1
	if allowed {
		value.tokens--
	}
	m.buckets[key] = value
	return allowed, nil
}

type MemoryIdempotencyStore struct {
	mu     sync.RWMutex
	values map[string]StoredResponse
	now    func() time.Time
}

func NewMemoryIdempotencyStore(now func() time.Time) *MemoryIdempotencyStore {
	return &MemoryIdempotencyStore{values: make(map[string]StoredResponse), now: now}
}

func (m *MemoryIdempotencyStore) Get(_ context.Context, key string) (StoredResponse, bool, error) {
	m.mu.RLock()
	value, ok := m.values[key]
	m.mu.RUnlock()
	if !ok {
		return StoredResponse{}, false, nil
	}
	if !value.ExpiresAt.After(m.now()) {
		m.mu.Lock()
		delete(m.values, key)
		m.mu.Unlock()
		return StoredResponse{}, false, nil
	}
	value.Header = value.Header.Clone()
	value.Body = append([]byte(nil), value.Body...)
	return value, true, nil
}

func (m *MemoryIdempotencyStore) Put(_ context.Context, key string, value StoredResponse, ttl time.Duration) error {
	value.Header = value.Header.Clone()
	value.Body = append([]byte(nil), value.Body...)
	value.ExpiresAt = m.now().Add(ttl)
	m.mu.Lock()
	m.values[key] = value
	m.mu.Unlock()
	return nil
}

File: src/rust/src/integrations.rs
use std::{collections::{BTreeMap, HashMap}, sync::Arc, time::Instant};

use async_trait::async_trait;
use tokio::sync::Mutex;

use crate::context::RequestContext;

#[derive(Debug, Clone)]
pub struct RequestMetadata {
    pub method: String,
    pub path: String,
    pub headers: BTreeMap<String, String>,
    pub remote_ip: Option<String>,
    pub content_length: Option<u64>,
}

#[derive(Debug, Clone)]
pub struct ResponseMetadata {
    pub status: u16,
    pub duration_ms: u64,
    pub response_bytes: Option<u64>,
}

#[derive(Debug, Clone, Default)]
pub struct AuthDecision {
    pub user_id: Option<String>,
    pub tenant_id: Option<String>,
    pub claims: BTreeMap<String, String>,
}

#[derive(Debug, Clone)]
pub struct IntegrationError {
    pub code: &'static str,
    pub message: String,
}

#[async_trait]
pub trait AuthVerifier: Send + Sync {
    async fn verify(&self, request: &RequestMetadata) -> Result<AuthDecision, IntegrationError>;
}

#[async_trait]
pub trait SyncObserver: Send + Sync {
    async fn observe(&self, context: &RequestContext, request: &RequestMetadata, response: &ResponseMetadata) -> Result<(), IntegrationError>;
}

#[async_trait]
pub trait TelemetrySink: Send + Sync {
    async fn request_started(&self, context: &RequestContext, request: &RequestMetadata);
    async fn request_finished(&self, context: &RequestContext, request: &RequestMetadata, response: &ResponseMetadata);
}

#[async_trait]
pub trait RateLimiter: Send + Sync {
    async fn allow(&self, key: &str, capacity: u32, refill_per_second: f64) -> bool;
}

#[derive(Default)]
pub struct AnonymousAuth;

#[async_trait]
impl AuthVerifier for AnonymousAuth {
    async fn verify(&self, _request: &RequestMetadata) -> Result<AuthDecision, IntegrationError> {
        Ok(AuthDecision::default())
    }
}

#[derive(Default)]
pub struct NoopSyncObserver;

#[async_trait]
impl SyncObserver for NoopSyncObserver {
    async fn observe(&self, _context: &RequestContext, _request: &RequestMetadata, _response: &ResponseMetadata) -> Result<(), IntegrationError> { Ok(()) }
}

#[derive(Default)]
pub struct TracingTelemetry;

#[async_trait]
impl TelemetrySink for TracingTelemetry {
    async fn request_started(&self, context: &RequestContext, request: &RequestMetadata) {
        tracing::info!(request_id = %context.request_id, trace_id = %context.trace_id, method = %request.method, path = %request.path, "request started");
    }
    async fn request_finished(&self, context: &RequestContext, _request: &RequestMetadata, response: &ResponseMetadata) {
        tracing::info!(request_id = %context.request_id, trace_id = %context.trace_id, status = response.status, duration_ms = response.duration_ms, "request finished");
    }
}

#[derive(Default)]
pub struct InMemoryTokenBucket {
    buckets: Mutex<HashMap<String, Bucket>>,
}

struct Bucket {
    tokens: f64,
    last_refill: Instant,
}

#[async_trait]
impl RateLimiter for InMemoryTokenBucket {
    async fn allow(&self, key: &str, capacity: u32, refill_per_second: f64) -> bool {
        let mut buckets = self.buckets.lock().await;
        let now = Instant::now();
        let bucket = buckets.entry(key.to_owned()).or_insert(Bucket { tokens: capacity as f64, last_refill: now });
        let elapsed = now.duration_since(bucket.last_refill).as_secs_f64();
        bucket.tokens = (bucket.tokens + elapsed * refill_per_second).min(capacity as f64);
        bucket.last_refill = now;
        if bucket.tokens >= 1.0 {
            bucket.tokens -= 1.0;
            true
        } else {
            false
        }
    }
}

pub type DynAuthVerifier = Arc<dyn AuthVerifier>;
pub type DynSyncObserver = Arc<dyn SyncObserver>;
pub type DynTelemetrySink = Arc<dyn TelemetrySink>;
pub type DynRateLimiter = Arc<dyn RateLimiter>;
